Components/Core/Timer.cpp

```diff
@@ -1,34 +1,226 @@
+#include "SystemDefines.hpp"
+#include "Timer.hpp"
+
 /**
  ******************************************************************************
  * File Name          : Timer.cpp
  * Description        : FreeRTOS Timer Wrapper
  ******************************************************************************
 */
-#include "SystemDefines.hpp"
-#include "Timer.hpp"
 
 /**
- * @brief Empty callback function, used internally for default polling timers
-*/
-void empty_callback(TimerHandle_t rtTimerHandle) {};
-
-/**
- * @brief Default constructor makes a timer that can only be polled for state
+ * Default constructor makes a timer that can only be polled for state
+ * Default behaviour : ->Autoreload is set to false (One shot Timer)
+ * 					   ->Timer Period is 1000ms
+ * 					   ->The Callback function simply changes state to COMPLETE and has no other functionality
 */
 Timer::Timer()
 {
 	// We make a timer named "Timer" with a callback function that does nothing, Autoreload false, and the default period of 1s.
 	// The timer ID is specified as (void *)this to provide a unique ID for each timer object - however this is not necessary for polling timers.
 	// The timer is created in the dormant state.
-	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, empty_callback);
+	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, CallbackFunction);
+	SOAR_ASSERT(rtTimerHandle, "Error Occurred, Timer not created");
+	timerState = UNINITIALIZED;
+}
+
+/**
+ * Constructor for callback enabled timer
+ * ! User has to add "Timer::CallbackFunction(rtTimerHandle);" in the callback function for accurate functioning of Timer States
+ * Default behaviour : ->Autoreload is set to false (One shot Timer)
+ * 					   ->Timer Period is 1000ms
+ * 					   ->The Callback function will be provided by the user while making sure to follow instruction above
+*/
+Timer::Timer(void (*TimerCallbackFunction_t)( TimerHandle_t xTimer ))
+{
+	rtTimerHandle = xTimerCreate("Timer", DEFAULT_TIMER_PERIOD, pdFALSE, (void *)this, TimerCallbackFunction_t);
+	SOAR_ASSERT(rtTimerHandle, "Error Occurred, Timer not created");
+	timerState = UNINITIALIZED;
+}
+
+/**
+ * @brief Default de-constructor makes a timer that can only be polled for state
+*/
+Timer::~Timer()
+{
+	if (xTimerDelete(rtTimerHandle, DEFAULT_TIMER_COMMAND_WAIT_PERIOD*2) == pdPASS) {
```

Comment:
If this fails, (else block) please print out something like "Warning, failed to delete timer"

Reply:
wouldn't this help for testing purposes only because we would never actually see this message anyways for actual flight

Reply:
Good point, depends on if we route traces over radio (which I want, but we'll have to see if the system can handle it)

I'd add it either way, since rtos things not deleting would be quite bad, it's more important than the "Timer deleted" message.

(when everything is running at the same time we won't always know exactly when a timer should be deleted or not, but we should know if it can't be deleted for any reason)

Reply:
yep makes sense. also might help if this flight log is actually stored. it might help for debugging after flight if we had any issues